### DISCUSSION TITLE
Variables missing in .mat file

### DISCUSSION THREAD

**Discussion by sbar123** [OpenModelica/OpenModelica#12147] [Q&A] [answered]
Hello

I am currently facing some issues with the data files written by openmodelica. When I run a simulation an read the .mat file  not all variables appear in the file. The variable name appears in the “name” section, but there are no corresponding values in the “data_2” section. All variables are visible in the OMEdit Variables browser. If I choose .csv a output file type, all variable appear in the file too. I can’t see a pattern with the variables that do not appear in the mat file. I have identically defined variable just with different names (both not protected or hideResult ). One appears the other not.. Can you help me?

Thanks you in advance!
Sebastian

I am simulating with OMedit 1.19.2 and tested also with 1.21.0

**Comment by casella** [comment]
@sbar123 please use 1.22.3 and the 1.23.0-dev nightly build and report if you still have the same issues

**Reply by sbar123** [reply]
In version 1.21, I export values for 317 variables (without any additional filters). However, in version 1.22.3, the exported values are for 249 variables (again, without any additional filters). This represents a reduction of approximately 20%. The same reduction also applies to the file size of the .mat file. I have trouble to get the nightly running I'll check as soon as i managed to run it!

**Reply by casella** [reply]
OK. Maybe some of those variables are protected? In that case, by default they are not saved, but you can set up the simulation output to include them.

**Reply by sbar123** [reply]
HI, no they are not protected!

**Reply by casella** [reply]
Can you post an example that reproduces the issue?

**Comment by hillbu1** [comment]
Hello,

I'm a colleague of sbar123 and working together with him on that topic.
The variable we are talking about are initialized inside a "sub-model" of a bigger system as

```
Modelica.Units.SI.Temperature A;
Modelica.Units.SI.Temperature B;
```

Then there are two sensors declared as protected:
```
  Modelica.Thermal.HeatTransfer.Sensors.TemperatureSensor sensor_A annotation(
    Placement(visible = true, transformation(extent = {{20, 40}, {40, 60}}, rotation = 0)));
  Modelica.Thermal.HeatTransfer.Sensors.TemperatureSensor sensor_B annotation(
    Placement(visible = true, transformation(extent = {{36, -34}, {56, -14}}, rotation = 0)));
```

In the equations section of that sub-model values are assigned to A and B:

```
A = sensor_A.T;
B = sensor_B.T;
```

In the end, only A is exported to the .mat file, B not. We do not understand this, as they are both handled in the same way.

**Comment by casella** [comment]
Maybe you can try to build a small model with these features that we can run?

**Comment by phannebohm** [comment]
Are some of those variables aliases? Search the [users guide](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/technical_details.html) for `dataInfo`.

**Reply by hillbu1** [reply]
I can find the names of the missing variables ***A*** in the mat_file.name category at positions 584, while my mat_file.data2 category only contains 249 data sets, i.e., 248 variables + time.
Looking into the mat_file.datainfo(584,2), it is [2]. Does that mean that the variable in mat_file.data2(:,2) - let's call it ***A<sub>alternative</sub>*** -  contains the same values and I could use these values instead? From the physical point of view it makes sense in our case.

Now if I repeat the same experiment but apply a VariableFilter filtering for ***A*** in the simulate command to reduce the file size of the output file, I get the variable ***A<sub>alternative</sub>*** still at index 2, while "A" itself is at index 340 without an own dataset. But mat_file.datainfo(340,2) is 2, still pointing to ***A<sub>alternative</sub>***.

So I understand that Modelica only exports one variable (to reduce file size) and links other variables to the exported one, if they are pyhsically identical in the model (e.g., directly conncected without any thermal resistance in between, like in our case). Even if I filter the variables for further reduction of file size, this link between the variables is still maintained and I get those variables that are identical and also the link of my desired variable to the identical one (having the dataset included). 
Did I get the concept right?
Then it would mean that I just learned about the alias variable and that my desired values are not missing in the output file. 👍 

Would there be any way to select a certain variable (that I'm actually interested in) as the primary to be shown in the results and take the others as the alias variable? Would make it more easy to filter the results afterwards (but actually can be overcome quite easily by some additional small post processing routine on the data).

**Reply by phannebohm** [reply, marked as answer]
Yes I think you get the idea. Let me clarify a few potential misconceptions though.

When the flat model contains equations of the form `a = b` or `a = -b` (which happens a lot when you use connectors), we collect alias sets. For example if we have `a = b` and `a = c`, we get the set `{a, b, c}`. For each alias set we keep one representative (e.g. `b`) from that set and replace all other variables (`a` and `c`) by that representative. We also remove all alias equations. With this Alias Replacement, as you noticed, we are able to reduce the size of the system significantly, which is good for storing results efficiently, but also for compiling and simulating the system faster.

The choice of representative is done internally by the compiler and you don't really have a way to influence that manually, I think. But you don't need to, because in your example the values in `data2(:,2)` __are the values of__ `A`, they are perhaps also the values of some other variable if they are in the same alias set, i.e. pointing to the same index in `dataInfo`. Don't rely on the index in `name` to correspond to the index in `data_1` or `data_2`! The mapping from indices in `name` to indices in `data_1` or `data_2` is stored in `dataInfo(:,2)`, so that level of indirection is always there.

FYI, in the new backend we are extending the notion of alias to the more general case `a = p*b + q` where `p` and `q` are constant expressions. Some day we may extend the *.mat format to also store those aliases efficiently.

**Reply by hillbu1** [reply]
> Don't rely on the index in name to correspond to the index in data_1 or data_2! The mapping from indices in name to indices in data_1 or data_2 is stored in dataInfo(:,2), so that level of indirection is always there.

This is good to know, so I'll always take the path via `dataInfo(:,2)` to get the correct indices of my values.

Thank you very much for your help and the latest additional information! I think the topic can be closed then. :-)
